### ISSUE TITLE
Alternator: measure and improve performance of long queries/scans

### ISSUE THREAD

**Issue by nyh** [scylladb/scylla#6278]
We do not yet have benchmarks measuring the full-table Scan or whole-partition Query operations, and we should. (for Query, there is also a reverse iteration - see issue #5153).

I think today, the performance of these scanning operations would be sub-optimal. The main problem is that these operations only read chunks of up a certain page size. While in CQL we have a mechanism of returning a "paging state" cookie and resume it later, to avoid some of the work when continuing the next page, in the DynamoDB API we lose this paging state cookie, and return only the current _position_ of the scan. Maybe we could keep a cache of paging states for the last stop positions (similarly to how in CQL we have a cache of those paging state cookies), so we can resume these scans.

**Comment by denesb** [comment]
> Maybe we could keep a cache of paging states for the last stop positions (similarly to how in CQL we have a cache of those paging state cookies), so we can resume these scans.

We don't have a cache of paging state cookies in CQL, the cookies are sent to the client who is responsible for re-sending them on the next page, much like how cookies stored in the browser work.

**Comment by nyh** [comment]
> We don't have a cache of paging state cookies in CQL, the cookies are sent to the client who is responsible for re-sending them on the next page, much like how cookies stored in the browser work.

Yes, the cookies are the **key** into the cache, not the value (the value is full-blown readers, with buffers, etc.). But Scylla still has this cache, which is what I meant when I mentioned a cache.
The problem in Alternator is that the DynamoDB API doesn't have the notion of these paging cookies - all it gives back when stopping a scan is the last position in the scan. So we probably need to have a cache whose key is the query parameters and the  position and - not a "cookie".

**Comment by denesb** [comment]
> > We don't have a cache of paging state cookies in CQL, the cookies are sent to the client who is responsible for re-sending them on the next page, much like how cookies stored in the browser work.
> 
> Yes, the cookies are the **key** into the cache, not the value (the value is full-blown readers, with buffers, etc.). But Scylla still has this cache, which is what I meant when I mentioned a cache.
> The problem in Alternator is that the DynamoDB API doesn't have the notion of these paging cookies - all it gives back when stopping a scan is the last position in the scan. So we probably need to have a cache whose key is the query parameters and the position and - not a "cookie".

I see. The key to the querier cache is a compound one, its components being:
* query uuid - a unique identifier for the query
* schema
* partition range
* partition slice

Out of these 4, 3 can be derived just from the query in question + the position to continue from. What remains is the query uuid. This uuid is currently generated by the pager on the first page, but this can be trivially modified to receive it from the outside. So if you have a deterministic and reliable way to map a query to one such uuid this is a solved problem.
Note that you also need to remember which replicas were used in the last page and send the next page request to the same set, otherwise you miss the cache readers.

**Comment by nyh** [comment]
We should consider how the feature requested here cooperates with various load-balancing options. If the scan request is sent by a load balancer to a different coordinator than the previous page of the scan, it will not be able to enjoy the cache I proposed. We should perhaps consider how this affect the design of Alternator load balancing (https://github.com/scylladb/alternator-load-balancing) and how it attempts (or doesn't attempt) to use the same node instead of skipping between different nodes on each request.
